File: app/src/main/java/com/eex/home/activity/login/RegisterActivity.java
package com.eex.home.activity.login;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.eex.R;
import com.eex.common.api.ApiFactory;
import com.eex.common.api.RxSchedulers;
import com.eex.common.base.BaseActivity;
import com.eex.common.base.Data;
import com.eex.common.util.CommonUtil;
import com.eex.common.util.TimeCount;
import com.eex.common.view.ComTitleBar;
import com.eex.home.weight.Utils;

import java.util.HashMap;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * _ooOoo_
 * o8888888o
 * 88" . "88
 * (| -_- |)
 * O\ = /O
 * ____/`---'\____
 * . ' \\| |// `.
 * / \\||| : |||// \
 * / _||||| -:- |||||- \
 * | | \\\ - /// | |
 * | \_| ''\---/'' | |
 * \ .-\__ `-` ___/-. /
 * ___`. .' /--.--\ `. . __
 * ."" '< `.___\_<|>_/___.' >'"".
 * | | : `- \`.;`\ _ /`;.`/ - ` : | |
 * \ \ `-. \_ __\ /__ _/ .-` / /
 * ======`-.____`-.___\_____/___.-`____.-'======
 * `=---='
 * <p>
 * .............................................
 * 佛祖保佑 永无BUG
 *
 * @ProjectName: OverThrow
 * @Package: com.overthrow.home.activity
 * @ClassName: RegisterActivity
 * @Description: java类作用描述
 * @Author: 胡成军
 * @CreateDate: 2019/6/6 16:06
 * @UpdateUser: 更新者
 * @UpdateDate: 2019/6/6 16:06
 * @UpdateRemark: 更新说明
 * @Version: 1.0
 * 邮箱注册
 */
public class RegisterActivity extends BaseActivity {


    /**
     * 头部
     */
    @BindView(R.id.comtitlebar)
    ComTitleBar comtitlebar;
    /**
     * 邮箱账号
     */
    @BindView(R.id.login_username_edit)
    EditText loginUsernameEdit;
    /**
     * 输入密码
     */
    @BindView(R.id.login_userpaswod_edit)
    EditText loginUserpaswodEdit;
    /**
     * 再次输入密码
     */
    @BindView(R.id.new_login_userpaswod_edit)
    EditText newLoginUserpaswodEdit;

    /**
     * 输入验证码
     */
    @BindView(R.id.login_userYZM_edit)
    EditText loginUserYZMEdit;
    /**
     * 获取短信
     */
    @BindView(R.id.btn)
    Button btn;
    /**
     * 推荐码
     */
    @BindView(R.id.login_userTJ_edit)
    EditText loginUserTJEdit;
    /**
     * 同意服务协议
     */
    @BindView(R.id.ck_register)
    CheckBox ckRegister;
    /**
     * 服务协议
     */
    @BindView(R.id.sever_textView)
    TextView severTextView;
    /**
     * 注册
     */
    @BindView(R.id.login_login_paword)
    Button loginLoginPaword;
    /**
     * 已有账号
     */
    @BindView(R.id.NOpaswd)
    TextView NOpaswd;
    /**
     * 手机注册
     */
    @BindView(R.id.putUser)
    TextView putUser;


    /**
     * 验证码时间
     */
    private TimeCount time;
    /**
     * 自定义字段 手机号码
     */
    private String Ecall;
    /**
     * 自定义字段 密码
     */
    private String password;
    /**
     * 自定义字段 确认密码
     */
    private String surePassword;
    /**
     * 自定义字段 短信验证码
     */
    private String verifyCode;
    /**
     * 自定义字段 推荐码
     */
    private String referralCode;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_register;
    }

    @Override
    protected void refreshData(Bundle savedInstanceState) {

        comtitlebar.setTitle(getActivity().getResources().getString(R.string.register));
    }

    @Override
    protected void initUiAndListener() {

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // TODO: add setContentView(...) invocation
        ButterKnife.bind(this);
    }

    @OnClick({R.id.comtitlebar, R.id.login_username_edit, R.id.login_userpaswod_edit, R.id.new_login_userpaswod_edit, R.id.login_userYZM_edit, R.id.btn, R.id.login_userTJ_edit, R.id.ck_register, R.id.sever_textView, R.id.login_login_paword, R.id.NOpaswd, R.id.putUser})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.comtitlebar:
                finish();
                break;
            //获取输入的邮箱账号
            case R.id.login_username_edit:
                Ecall = loginUsernameEdit.getText().toString().trim();
                break;

            //密码
            case R.id.login_userpaswod_edit:
                password = loginUserpaswodEdit.getText().toString().trim();
                break;
            //确认密码
            case R.id.new_login_userpaswod_edit:
                surePassword = newLoginUserpaswodEdit.getText().toString().trim();
                break;
            //验证码
            case R.id.login_userYZM_edit:
                verifyCode = loginUserYZMEdit.getText().toString().trim();
                break;
            //获取验证码
            case R.id.btn:

                Ecall = loginUsernameEdit.getText().toString().trim();
                //判断邮箱号是否为null
                if (TextUtils.isEmpty(Ecall)) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.please_input_phone));
                    return;
                }
                VerifyCode(loginUsernameEdit.getText().toString().trim());
                break;
            //推荐码
            case R.id.login_userTJ_edit:
                referralCode = loginUserTJEdit.getText().toString().trim();
                break;
            case R.id.ck_register:
                break;
            //服务协议
            case R.id.sever_textView:
                intent.setClass(RegisterActivity.this, SeverActivity.class);
                startActivity(intent);
                break;
            //注册
            case R.id.login_login_paword:
                //获取邮箱账号
                Ecall = loginUsernameEdit.getText().toString().trim();
                //密码
                password = loginUserpaswodEdit.getText().toString().trim();
                //确认密码
                surePassword = newLoginUserpaswodEdit.getText().toString().trim();
                //验证码
                verifyCode = loginUserYZMEdit.getText().toString().trim();
                //推荐码
                referralCode = loginUserTJEdit.getText().toString().trim();


                //判断邮箱账号是否为null
                if (TextUtils.isEmpty(Ecall)) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.please_input_phone));
                    return;
                }
                //判断密码是否为null
                if (TextUtils.isEmpty(password)) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.please_input_need_password));
                    return;
                }
                //判断确认密码是否为null
                if (TextUtils.isEmpty(surePassword)) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.please_input_confirm_password));
                    return;
                }
                //判断短信验证码是否为null
                if (TextUtils.isEmpty(verifyCode)) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.login_yzm));
                    return;
                }


                //是否同意了服务协议
                if (ckRegister.isChecked() == false) {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.please_check_agreement));
                    return;
                }

                if (password.length() < 6) {
                    Toast.makeText(this, R.string.please_input_need_password, Toast.LENGTH_SHORT).show();
                    return;
                }


                //判断输入密码和输入的确认密码一样
                if (password.equals(surePassword)) {
                    registersure(Ecall, password, verifyCode, referralCode);
                } else {
                    CommonUtil.showSingleToast(getActivity().getResources().getString(R.string.two_passwords_not_same));

                }

                break;

            //已有账号，返回登录
            case R.id.NOpaswd:
                intent.setClass(RegisterActivity.this, LoginActivity.class);
                startActivity(intent);
                finish();

                break;

            //邮箱注册
            case R.id.putUser:
                intent.setClass(RegisterActivity.this, PhoneRegisterActivity.class);
                startActivity(intent);
                finish();
                break;


            default:
                break;
        }
    }


    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN
                && event.getKeyCode() == KeyEvent.KEYCODE_BACK) {
            finish();
            return true;
        }
        return super.dispatchKeyEvent(event);
    }

    /**
     * 获取邮箱验证码
     *
     * @param ecall
     */
    @SuppressLint("CheckResult")
    private void VerifyCode(String ecall) {

        HashMap<String, String> hashMap = new HashMap<String, String>();

        hashMap.put("sendType", "4");
        hashMap.put("email", loginUsernameEdit.getText().toString().trim() + "");
        Log.e("TAG", "VerifyCode: " + loginUsernameEdit.getText().toString().trim());
        ApiFactory.getInstance()
                .email1(hashMap)
                .compose(RxSchedulers.<Data>io_main())
                .subscribe(data -> {
                    if (data.isStauts() == true) {
                        //构造CountDownTimer对象
                        time = new TimeCount(RegisterActivity.this, btn, getActivity().getResources().getString(R.string.regain_code), 60000, 1000);
                        time.start();
                    } else {
                        t(data.getMsg());
                    }
                }, throwable -> {

                });
    }


    /**
     * 邮箱注册
     *
     * @param ecall
     * @param password
     * @param verifyCode
     * @param referralCode
     */
    @SuppressLint("CheckResult")
    private void registersure(String ecall, String password, String verifyCode, String referralCode) {


        HashMap<String, String> hashMap = new HashMap<String, String>();

        if (referralCode.equals("")) {

        } else {
            // 推荐码
            hashMap.put("invateCode", referralCode + "");
        }
        // 电话
        hashMap.put("username", ecall + "");
        //密码
        hashMap.put("password", Utils.md5(password + "hello, moto"));
        // 验证码

        hashMap.put("code", verifyCode);

        ApiFactory.getInstance()
                .regist(hashMap)
                .compose(RxSchedulers.<Data>io_main())
                .subscribe(data -> {
                    if (data.getCode() == 200) {

                        Intent intent = new Intent();
                        intent.setClass(RegisterActivity.this, LoginActivity.class);
                        startActivity(intent);
                        finish();
                    } else {
                        t(getActivity().getResources().getString(R.string.registerNo));

                    }

                }, throwable -> {
                    t(getActivity().getResources().getString(R.string.wangl));
                });

    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }
}

File: app/src/main/java/com/eex/mvp/usercenter/register/RegisterActivity.kt
package com.eex.mvp.usercenter.register

import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Base64
import android.view.View
import com.eex.R
import com.eex.common.util.CommonUtil
import com.eex.domin.entity.register.Register
import com.eex.home.weight.Utils
import com.eex.mvp.MVPBaseActivity
import com.eex.navigation.Navigator
import kotlinx.android.synthetic.main.re_activity_register.*

/**
 * @Description:注册（手机注册和邮箱注册综合）
 * @Author: xq
 * @CreateDate: 2019/11/28 16:45
 */
class RegisterActivity : MVPBaseActivity<Register, RegisterContract.View, RegisterPresenter>(), RegisterContract.View {

    override val layoutId: Int = R.layout.re_activity_register

    private var type = 1 //1手机注册 2邮箱
    private var register: Register? = null

    override fun getImageCodeSuccess(register: Register) {
        this.register = register
        //将Base64编码字符串解码成Bitmap
        val decodedString = Base64.decode(register.imageCode.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1], Base64.DEFAULT)
        val decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.size)
        //设置ImageView图片
        graphics_image.setImageBitmap(decodedByte)
    }

    override fun getCodeSuccess(register: Register) {
        showToast(register.msg)
        presenter.openTimer(this, btn_YZM, getString(R.string.regain_code))
    }

    override fun registerSuccess(register: Register) {
        showToast(register.msg)
        Navigator.toLoginActivity(this@RegisterActivity)
//        finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initView()
    }

    private fun initView() {

        register_back.setOnClickListener {
            CommonUtil.hideKeyboard(this)
            finish()
        }

        presenter.getImageCode()
        graphics_linear.visibility = if (type == 1) View.VISIBLE else View.GONE

        ll_Phone.setOnClickListener {
            Navigator.toPhoneListActivity(this@RegisterActivity,2000)
        }

        login_username_edit.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {

            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (type == 1) {
                    if (login_username_edit.text.toString().trim { it <= ' ' }.length >= 4 && Utils.isInteger(login_username_edit.text.toString().trim { it <= ' ' })) {
                        ll_Phone.visibility = View.VISIBLE
                    } else {
                        ll_Phone.visibility = View.GONE
                    }
                }
            }

        })

        btn_YZM.setOnClickListener {
            val input = login_username_edit.text.toString().trim { it <= ' ' }
            if (TextUtils.isEmpty(input)) {
                showToast(if (type == 1) R.string.please_input_phone else R.string.please_input_email)
                return@setOnClickListener
            }
            when (type) {
                1 -> {
                    if (register == null) {
                        showToast(R.string.getting_image_code_error)
                        return@setOnClickListener
                    }
                    val imageCode = graphics_YZM_edit.text.toString().trim { it <= ' ' }
                    if (TextUtils.isEmpty(imageCode)) {
                        showToast(R.string.please_input_image_code)
                        return@setOnClickListener
                    }
                    presenter.getPCode(tx_name_nuber.text.toString() + " " + input, imageCode, register!!.imageKey)
                }
                else -> {
                    presenter.getECode(input)
                }
            }
        }
        sever_textView.setOnClickListener {
            //跳转查看用户协议界面
            Navigator.toSeverActivity(this@RegisterActivity)
        }

        login_login_paword.setOnClickListener {
            //获取注册账号
            val input = login_username_edit.text.toString().trim { it <= ' ' }
            //密码
            val password = login_userpaswod_edit.text.toString().trim { it <= ' ' }
            //确认密码
            val surePassword = new_login_userpaswod_edit.text.toString().trim { it <= ' ' }
            //验证码
            val verifyCode = login_userYZM_edit.text.toString().trim { it <= ' ' }
            //推荐码
            val referralCode = login_userTJ_edit.text.toString().trim { it <= ' ' }
            //判断手机号码是否为null
            if (TextUtils.isEmpty(input)) {
                showToast(if (type == 1) R.string.please_input_phone else R.string.please_input_email)
                return@setOnClickListener
            }
            //判断密码是否为null
            if (TextUtils.isEmpty(password) || password.length < 6) {
                showToast(R.string.please_input_need_password)
                return@setOnClickListener
            }
            //判断确认密码是否为null
            if (TextUtils.isEmpty(surePassword)) {
                showToast(R.string.please_input_confirm_password)
                return@setOnClickListener
            }
            //判断短信验证码是否为null
            if (TextUtils.isEmpty(verifyCode)) {
                showToast(R.string.login_yzm)
                return@setOnClickListener
            }

            //是否同意了服务协议
            if (!ck_register.isChecked) {
                showToast(R.string.please_check_agreement)
                return@setOnClickListener
            }

            //判断输入密码和输入的确认密码一样
            if (password != surePassword) {
                showToast(R.string.two_passwords_not_same)
                return@setOnClickListener
            }

            when (type) {
                1 -> presenter.register(tx_name_nuber.text.toString() + " " + input, Utils.md5(password + "hello, moto"), verifyCode, referralCode)
                2 -> presenter.register(input, Utils.md5(password + "hello, moto"), verifyCode, referralCode)
            }

        }

        putUser.setOnClickListener {
            if (type == 1) {
                type = 2
                login_username_edit.hint = getString(R.string.please_input_email)
                putUser.text = getString(R.string.phone_register)
                graphics_linear.visibility = View.GONE
                ll_Phone.visibility = View.GONE
            } else {
                type = 1
                login_username_edit.hint = getString(R.string.please_input_phone)
                putUser.text = getString(R.string.email_register)
                graphics_linear.visibility = View.VISIBLE
                if (login_username_edit.text.toString().trim { it <= ' ' }.length >= 4 && Utils.isInteger(login_username_edit.text.toString().trim { it <= ' ' })) {
                    ll_Phone.visibility = View.VISIBLE
                } else {
                    ll_Phone.visibility = View.GONE
                }
            }
        }
        NOpaswd.setOnClickListener {
            finish()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 2000) {
            if (data != null) {
                val name = data.getStringExtra("name")
                val code = data.getStringExtra("code")
                guojia_name.text = name
                tx_name_nuber.text = code
            }
        }
    }
}